django/template/base.py

```diff
@@ -187,12 +187,13 @@ def __init__(self, template_string, origin=None, name=None, engine=None):
         if engine is None:
             from .engine import Engine
             engine = Engine.get_default()
-        if engine.debug and origin is None:
+        if origin is None:
             origin = StringOrigin(template_string)
-        self.nodelist = engine.compile_string(template_string, origin)
         self.name = name
         self.origin = origin
         self.engine = engine
+        self.source = template_string
+        self.nodelist = engine.compile_string(self)
```

Comment:
I'm reluctant to pass a `Template` to the `Engine` because that creates a circular relation between the two classes.

If you keep this design, you must at least rename `compile_string`, because it no longer accepts a string in argument. In fact I would add a new method and deprecate `compile_string`.